Hosting Dashboard: Site PHP and Web logs - introduce infinite scroll

coder-karen opened the pull request:
Part of DOTCOM-14330

## Proposed Changes

* This PR implements infinite scroll on the Hosting Dashboard Site log pages (PHP and Web server logs, not Activity).
* There is some customization (manually including infinite scroll capabilities) due to constraints (we can't pass the infinite scroll handler via `paginationInfo` as we don't want an infinite scroll toggle in the View settings).
* We're also including a back to top button.


## Why are these changes being made?

* This has been implemented because pagination is currently partially hidden on these pages (it is possible to navigate forward and back, but not to select a page number - just as with the pagination on the existing dashboard logs pages), as page selection won't work. Because of the likelihood of a large number of logs in any specified date range we're not using the Dataviews `filterSortAndPaginate` function here. This function works client-side, however fetching “all” to enable client-side filtering would be expensive. We implemented a server-side approach.
* Since we were using a hack for pagination, infinite scroll was recommended as a better solution.

## Testing Instructions

* Open up the site logs page at `http://calypso.localhost:3000/v2/sites/your-atomic-site/logs`.
* Visit the 'Web server' tab for a site that has existed for more than a day in order to have a larger number of entries showing.
* With the current default date range preset, scroll to the bottom of the page. More entries should load within the dataviews container.
* If this takes awhile to load, you should see a spinner, until the next batch of results appear.
* You should also see a scroll-to-top icon which can be used to take you to the top of the container again.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

coder-karen committed:
Introduce infinite scroll for the Hosting Dashboard PHP and Web logs pages

coder-karen committed:
Ensure infinite scroll more consistently loads by preventing dead ends

coder-karen committed:
Clear previous infinite query data when sort or filter changes to prevent empty results

coder-karen committed:
Merge branch 'trunk' into add/hosting-dashboard-site-logs-infinite-scroll

arthur791004 commented on packages/api-queries/src/site-logs.ts:
```diff
@@ -1,11 +1,20 @@
+		queryKey: [ 'site', siteId, 'logs', 'infinite', params ],
+		queryFn: ( { pageParam }: { pageParam: string | null } ) =>
+			fetchSiteLogs( siteId, params, pageParam ?? undefined ),
+		initialPageParam: null as string | null,
```
Should it be `undefined` as the `getNextPageParam` returns string or undefined instead of null?

coder-karen replied:
Yes good catch.

arthur791004 reviewed (commented):
Would it be better to use the infinite scroll of the DataViews by setting the max height? The current way is not easy to scroll back to switch the date and tab 😅

arthur791004 commented on client/dashboard/sites/logs/dataviews/index.tsx:
```diff
@@ -71,71 +85,110 @@ function SiteLogsDataViews( {
-	const { data: siteLogs, isFetching } = useQuery( siteLogsQuery( site.ID, params, scrollId ) );
+	const { data, isLoading, isFetching, isFetchingNextPage, fetchNextPage, hasNextPage } =
+		useInfiniteQuery( {
+			...siteLogsInfiniteQuery( site.ID, params ),
```
Looks like we don't need to use `...` here as we didn't add any property.

arthur791004 commented on client/dashboard/sites/logs/dataviews/index.tsx:
```diff
@@ -71,71 +85,110 @@ function SiteLogsDataViews( {
+	}, [ infiniteScrollHandler ] );
 
 	const fields = useFields(
 		timezoneString ? { logType, timezoneString, gmtOffset } : { logType, gmtOffset }
```
It seems the undefined case of the `timezoneString` has been handled by `useFields` so we can always pass `{ logType, timezoneString, gmtOffset }` 🤔

coder-karen replied:
Oh yes, good catch. Though not added in this PR, I've modified it here now.

arthur791004 commented on client/dashboard/sites/logs/utils.ts:
```diff
@@ -149,3 +149,42 @@ export function getDateTimeLabel( {
+		items.forEach( ( php, index ) => {
+			out.push( {
+				...php,
+				id: `${ php.timestamp }|${ php.file }|${ String( php.line ) }|${ suffix }|${ String(
```
Could we change to use the array.join here? Besides, I'm not sure the meaning of the ID here, can we simply generate the random 16chars string for it on the API response 🤔

coder-karen replied:
Wouldn't deterministic keys be better, as random would change per fetch and thus not be as stable?

As for adding the IDs on the API response, after moving ID creation into fetch, some rows were now sharing the same ID across pages, so keys aren’t unique and I was seeing React warnings.

I've left the changes in the UI but used join instead (and also added `getItemId` within `DataViews` as that was missing).

coder-karen committed:
Clean-up based on reviews

coder-karen committed:
Add back to top button and load more button

coder-karen commented:
@arthur791004 thanks for your reviews!

> Would it be better to use the infinite scroll of the DataViews by setting the max height? The current way is not easy to scroll back to switch the date and tab 😅

I did have that initially, but had header display issues as a result. What was happening is that using` DataViews.Layout` results in the DataViews table header (with the view settings icon, as well as the logs downloader and auto-refresh toggle) being removed (not just hidden). What I then later discovered is that if you pass a child prop to the DataViews component, it enters free composition mode (hence removing the headers): https://developer.wordpress.org/block-editor/reference-guides/packages/packages-dataviews/#free-composition

I began working on implementing this, but another hurdle is that `paginationInfo.infiniteScrollHandler` is only used by DataViews to control the toggle in ViewConfig from what I understand. Removing it stops DataViews from thinking infinite scroll is available, so the toggle disappears (and infinite scroll no longer works). But we can get around that with the `IntersectionObserver`.

So the resulting code is already a lot more than before, going beyond the default implementation, and at the point I switched course there were still styling issues thanks to the additional parent needed to wrap around the DataViews component.

What I have done instead for a new commit is add a Back to Top button. I also added a Load More button as it is somewhat simpler in implementation. I haven't updated the PR description just yet though.

I'm still thinking the current approach is the least complex (with or without a Load More button, but perhaps adding a Back to Top button).

That said, we're only implementing infinite scroll because it wasn't clear if the pagination customization approach was technically allowed (hiding the option to select the page). So if this will also work out to be too much customization, perhaps we leave the pagination approach as is 🤔

arthur791004 commented:
The newly added buttons look inconsistent with other DataViews 😅 And the "Back to Top" doesn't work on my end.

arthur791004 commented:
I meant something like https://github.com/Automattic/wp-calypso/pull/106093

arthur791004 commented:
Oh yes, I didn't notice that. It doesn't seem to have config to turn it off 🥲

arthur791004 committed:
Use built-in infiniteScrollHandler (#106093)

coder-karen committed:
Manually add infinite scroll to prevent toggle to disable showing, and add scroll to top button

arthur791004 commented on client/dashboard/sites/logs/utils.ts:
```diff
@@ -2,7 +2,21 @@ import { dateI18n } from '@wordpress/date';
+/**
+ * Simple throttle function to limit how often a function can be called.
+ */
+export function throttle( fn: () => void, limit: number ) {
```
we can import it from `@wordpress/compose`

coder-karen replied:
Oh! Thanks for spotting that.

arthur791004 commented on client/dashboard/sites/logs/dataviews/style.scss:
```diff
@@ -47,3 +47,22 @@
+	inset-block-end: 2rem;
+	inset-inline-end: 2rem;
+	background: var( --color-primary );
+	color: var( --studio-white );
```
We should not use these color variables. They come from the color schemes and are not introduced in HD v2.

coder-karen replied:
Thanks, removed those now.

arthur791004 commented on client/dashboard/sites/logs/dataviews/style.scss:
```diff
@@ -47,3 +47,22 @@
+
+/* Scroll to top button */
+.site-logs-card .site-logs-scroll-to-top {
+	position: fixed;
```
Would it be better to stick to the bottom of the DataViews?

```css
	position: absolute;
	right: 8px;
	bottom: 8px;
	border-radius: 50%;
	box-shadow: 0 2px 6px rgba( 0, 0, 0, 0.3 );
```

coder-karen replied:
Can do, yes -- moved it over.

arthur791004 commented on client/dashboard/sites/logs/dataviews/index.tsx:
```diff
@@ -259,6 +320,15 @@ function SiteLogsDataViews( {
+			{ showScrollTop && (
+				<Button
+					variant="tertiary"
+					className="site-logs-scroll-to-top"
```
It would be better to assign the icon with default variant for the icon button

```js
__next40pxDefaultSize
size="compact"
icon={ arrowUp }
iconSize={ 24 }
```

coder-karen replied:
Got it. Only change I didn't implement was `__next40pxDefaultSize` as I didn't see how it was relevant. Are you seeing a console warning without this?

arthur791004 replied:
It may be fine 🙂

arthur791004 commented on client/dashboard/sites/logs/dataviews/index.tsx:
```diff
@@ -230,14 +246,58 @@ function SiteLogsDataViews( {
-					data={ logs as PHPLog[] }
-					isLoading={ isFetching }
+					data={ phpLogs }
+					isLoading={ isLoading }
```
Looks like we should keep using `isFetching` as the loading state didn't work correctly.

coder-karen replied:
Good spot, thanks.

coder-karen committed:
Use wordpress/compose throttle function

coder-karen committed:
Icon style changes

coder-karen committed:
Switch to isFetching instead of isLoading for loading state

arthur791004 commented on client/dashboard/sites/logs/dataviews/style.scss:
```diff
@@ -47,3 +47,16 @@
+	border-radius: 50%;
+	box-shadow: 0 2px 6px rgba( 0, 0, 0, 0.3 );
+	color: var( --wp-admin-theme-color );
+	inline-size: 34px;
```
Is `inline-size` required 🤔

coder-karen replied:
The circle appeared a little too narrow to my eyes without it, but it may just be me. It's not much of a difference. I can remove it for simplicity.

arthur791004 commented on client/dashboard/sites/logs/dataviews/style.scss:
```diff
@@ -47,3 +47,16 @@
+	border: none;
+	border-radius: 50%;
+	box-shadow: 0 2px 6px rgba( 0, 0, 0, 0.3 );
+	color: var( --wp-admin-theme-color );
```
I'm not sure if it's worth customizing the color 🤔

coder-karen replied:
Possibly not. You could argue that it stands out more as a link this way, but the same coloring shows on hover, and the arrow also helps indicate it's functionality. I'll remove it, can always update it in the future.

arthur791004 reviewed (approved):
Works as expected! Left some minor feedback 🙂

coder-karen committed:
Simplify CSS

coder-karen committed:
Simplify CSS

coder-karen committed:
Simplify CSS

coder-karen committed:
Merge branch 'trunk' into add/hosting-dashboard-site-logs-infinite-scroll
